src/main/scala/is/hail/variant/VariantSampleMatrix.scala

```diff
@@ -40,81 +42,95 @@ object VariantSampleMatrix {
-    def doRead[T](genotypeImporter: ((Variant, Row) => Iterable[T], ClassTag[T])): VariantSampleMatrix[T] = {
-      val (importGenotypes, tct) = genotypeImporter
+    val vaImporter = SparkAnnotationImpex.annotationImporter(metadata.vaSignature)
 
-      def importVa(r: Row): Annotation =
-        if (vaRequiresConversion) SparkAnnotationImpex.importAnnotation(r.get(1), vaSignature) else r.get(1)
+    if (isGenericGenotype) {
+      implicit val kOk = vSignature.orderedKey
 
-      def importRow(r: Row) = {
-        val v = r.getVariant(0)
-        (v, (importVa(r),
-          if (dropSamples)
-            Iterable.empty[T]
-          else
-            importGenotypes(v, r)))
-      }
+      val vImporter = SparkAnnotationImpex.annotationImporter(metadata.vSignature)
+      val gImporter = SparkAnnotationImpex.annotationImporter(metadata.genotypeSignature)
 
       val orderedRDD =
         if (dropVariants)
-          OrderedRDD.empty[Locus, Variant, (Annotation, Iterable[T])](sc)
+          OrderedRDD.empty[Annotation, Annotation, (Annotation, Iterable[Annotation])](sc)
         else {
+          def importRow(r: Row): (Annotation, (Annotation, Iterable[Annotation])) = {
+            val v = vImporter(r.get(0))
+            (v, (vaImporter(r.get(1)),
+              if (dropSamples)
+                Iterable.empty[Annotation]
+              else
+                r.getSeq[Any](2).lazyMap(g => gImporter(g))))
+          }
+
+          val partitioner: OrderedPartitioner[Annotation, Annotation] =
+            try {
+              val jv = hConf.readFile(dirname + "/partitioner.json.gz")(JsonMethods.parse(_))
+              implicit val pkjr = vSignature.partitionKey.jsonReader
+              jv.fromJSON[OrderedPartitioner[Annotation, Annotation]]
+            } catch {
+              case _: FileNotFoundException =>
+                fatal("missing partitioner.json.gz when loading VDS, create with HailContext.write_partitioning.")
+            }
+
           val columns = someIf(dropSamples, Array("variant", "annotations"))
-          OrderedRDD[Locus, Variant, (Annotation, Iterable[T])](sqlContext.readParquetSorted(parquetFile, columns).map(importRow), partitioner)
+          OrderedRDD[Annotation, Annotation, (Annotation, Iterable[Annotation])](
+            sqlContext.readParquetSorted(parquetFile, columns).map(importRow), partitioner)
         }
 
-      new VariantSampleMatrix[T](hc, metadata, localValue, orderedRDD)(tct)
-    }
+      new VariantSampleMatrix[Annotation, Annotation, Annotation](hc, metadata, localValue, orderedRDD)
+    } else {
+      assert(vSignature == TVariant)
+      val orderedRDD = if (dropVariants)
+        OrderedRDD.empty[Locus, Variant, (Annotation, Iterable[Genotype])](sc)
+      else {
+        def importRow(r: Row): (Variant, (Annotation, Iterable[Genotype])) = {
+          val v = r.getVariant(0)
+          (v, (vaImporter(r.get(1)),
+            if (dropSamples)
+              Iterable.empty[Genotype]
```

Comment:
This should use `GenotypeStream.empty()` given my discovery about the cost of not doing anything :man_shrugging: